File: leetcode/src/BinaryTreePreorderTraversal.java
import java.util.ArrayList;
import java.util.List;

public class BinaryTreePreorderTraversal extends BaseTreeNode {
    public static void main(String[] args) {
        BinaryTreePreorderTraversal instance = new BinaryTreePreorderTraversal();
        System.out.println(instance.preorderTraversal(generateTreeNode(new Integer[]{1, null, 2, 3})));
    }
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<Integer>();
        preorderTraversal(root, result);
        return result;
    }
    private void preorderTraversal(TreeNode node, List<Integer> previousResult) {
        if (node != null) {
            previousResult.add(node.val);
            preorderTraversal(node.left, previousResult);
            preorderTraversal(node.right, previousResult);
        }
    }
}

File: leetcode/src/python/binary_tree_preorder_traversal.py
from typing import Optional
from typing import List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class BinaryTreePreorderTraversal:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = []
        if root is not None:
            result.append(root.val)
            result += self.preorderTraversal(root.left)
            result += self.preorderTraversal(root.right)
        return result